File: fitness_core/macros.py
"""Macro split logic. Calorie math: protein/carb 4 kcal/g, fat 9 kcal/g."""

KCAL_PER_G = {"protein": 4, "carb": 4, "fat": 9}


def macros_g_per_kg(calories: float, kg: float,
                    protein_g_per_kg: float, fat_g_per_kg: float) -> dict:
    protein_g = protein_g_per_kg * kg
    fat_g = fat_g_per_kg * kg
    remaining = calories - protein_g * 4 - fat_g * 9
    carb_g = max(0.0, remaining / 4)
    return {
        "protein_g": round(protein_g, 1),
        "fat_g": round(fat_g, 1),
        "carb_g": round(carb_g, 1),
        "calories": round(protein_g * 4 + fat_g * 9 + carb_g * 4, 0),
    }

File: packages/core/src/math/macros.ts
import { roundTo } from "./stats.js";

export const KCAL_PER_G = { protein: 4, carb: 4, fat: 9 } as const;

export interface MacroSplit {
  protein_g: number;
  fat_g: number;
  carb_g: number;
  calories: number;
}

/** Macro split (grams): protein and fat set by g/kg bodyweight, carbs fill the calorie budget. */
export function macrosGPerKg(
  calories: number,
  kg: number,
  proteinGPerKg: number,
  fatGPerKg: number,
): MacroSplit {
  const { protein: p, carb: c, fat: f } = KCAL_PER_G;
  const proteinG = proteinGPerKg * kg;
  const fatG = fatGPerKg * kg;
  const carbG = Math.max(0, (calories - proteinG * p - fatG * f) / c);
  return {
    protein_g: roundTo(proteinG, 1),
    fat_g: roundTo(fatG, 1),
    carb_g: roundTo(carbG, 1),
    calories: roundTo(proteinG * p + fatG * f + carbG * c, 0),
  };
}
